src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+use rustc::ty::{TyKind, TypeAndMut};
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    // pthread_mutexattr_t is either 4 or 8 bytes, depending on the platform
+    // memory layout: store an i32 in the first four bytes equal to the
+    // corresponding libc mutex kind constant (i.e. PTHREAD_MUTEX_NORMAL)
+
+    fn pthread_mutexattr_init(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr_place = this.deref_operand(attr_op)?;
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+        let default_kind = this.eval_libc("PTHREAD_MUTEX_DEFAULT")?;
+        this.write_scalar(default_kind, kind_place.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_settype(
+        &mut self,
+        attr_op: OpTy<'tcx, Tag>,
+        kind_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let kind = this.read_scalar(kind_op)?.not_undef()?;
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? ||
+                kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")? ||
+                kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
+            let attr_place = this.deref_operand(attr_op)?;
+            let i32_layout = this.layout_of(this.tcx.types.i32)?;
+            let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+            this.write_scalar(kind, kind_place.into())?;
+        } else {
+            let einval = this.eval_libc_i32("EINVAL")?;
+            return Ok(einval);
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_destroy(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr_place = this.deref_operand(attr_op)?;
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+        this.write_scalar(ScalarMaybeUndef::Undef, kind_place.into())?;
```

Comment:
I think this is the first time I see a deliberate write of `Undef` in Miri. ;) Makes sense though.

Reply:
We have an open issue to make `Operand::Move` invalidate the value source

Reply:
Yeah we do, though it is also unclear if that is really what we want the behavior of the Rust Abstract Machine to be.